### PR TITLE
Added logging unsupported content type

### PR DESCRIPTION
**Pull request by oleg-smith**

### CODE REVIEW

**Comment by jlprat** [comment]
OK TO TEST

**Comment by oleg-smith** [comment]
@jlprat Do you think the binary compatibility problem can be ignored?

**Comment by jlprat** [comment]
Sorry I'm not close to a computer now, I will take a look later, but as a general rule binary compatible problems can only be "ignored" if the class is labeled as `@InternalApi` or `@ApiMayChange`. There are other cases when adding methods though.

**Comment by jlprat** [comment]
Hi @oleg-smith 
Sorry for the delay, but I was away from a computer. IMO this change is binary incompatible the way it is. For example, the Exception `UnsupportedContentTypeException` now has a different constructor and any 3rd party library built upon previous versions of Akka HTTP will fail if we introduce this change.
I would try to implement this change in a compatible way, for example, offering 2 constructors and some default value for `contentType`.

**Comment by jlprat** [comment]
One extra thing, is there an issue for this PR?

**Comment by oleg-smith** [comment]
Yes, #1691

**Comment by oleg-smith** [comment]
Will implement it in compatible way.

**Comment by jrudolph** [comment]
PLS BUILD

**Comment by oleg-smith** [comment]
@jlprat @jrudolph can this incompatibilty be ignored?
`[error]  * the type hierarchy of object akka.http.scaladsl.server.UnsupportedRequestContentTypeRejection is different in current version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("akka.http.scaladsl.server.UnsupportedRequestContentTypeRejection$")`

**Comment by oleg-smith** [comment]
@jlprat I have finished task.

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala [hunk 1]
```diff
@@ -154,7 +148,40 @@ object Unmarshaller
         s"Right failure: ${right.getMessage}, " +
         s"Left failure: ${left.getMessage}")
 
+  /**
+   * Signals that unmarshalling failed because the entity content-type did not match one of the supported ranges.
+   * This error cannot be thrown by custom code, you need to use the `forContentTypes` modifier on a base
+   * [[akka.http.scaladsl.unmarshalling.Unmarshaller]] instead.
+   */
+  final case class UnsupportedContentTypeException(
```
Comment (raboof) [inline akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: Do we still want this to be a case class?

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala [hunk 2]
```diff
@@ -154,7 +148,40 @@ object Unmarshaller
...
+    supported:   Set[ContentTypeRange],
+    contentType: Option[ContentType]   = None)
+    extends RuntimeException(supported.mkString(
+      s"Unsupported Content-Type [$contentType], supported: ", ", ", "")) {
+
+    //constructors and copy methods added to cover binary compatibility
+    def this(supported: Set[ContentTypeRange]) = this(supported, None)
```
Comment (raboof) [inline akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: do we want to mark these as deprecated as they're only here for binary compat?

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala [hunk 3]
```diff
@@ -154,7 +148,40 @@ object Unmarshaller
...
         s"Left failure: ${left.getMessage}")
 
+  /**
+   * Signals that unmarshalling failed because the entity content-type did not match one of the supported ranges.
+   * This error cannot be thrown by custom code, you need to use the `forContentTypes` modifier on a base
+   * [[akka.http.scaladsl.unmarshalling.Unmarshaller]] instead.
+   */
+  final case class UnsupportedContentTypeException(
+    supported:   Set[ContentTypeRange],
+    contentType: Option[ContentType]   = None)
```
Comment (raboof) [inline akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: Is the only reason to have this as an option binary compatibility? That's kind of unfortunate... perhaps we should consider introducing a singleton `ContentType` to use in that case?

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 1 above)
Comment (oleg-smith) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: Yes, as moving it to non-case class could cause more binary incompatibility issues

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 3 above)
Comment (oleg-smith) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: no, the reason is `Content-Type` is not always available

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 2 above)
Comment (oleg-smith) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: ok

**Comment by oleg-smith** [comment]
@jlprat @raboof can merge?

**Comment by jlprat** [comment]
Hi @oleg-smith!
Thanks for your contribution and welcome to Akka HTTP! In order to be merged, we need first 2 people approving the changes. After this, the PR will be merged.
I'll try to take a look later today to this PR.

**Comment by jrudolph** [comment]
I think we haven't pulled it off so far to extend a case class in such a way. I think we need to be quite sure that we don't introduce binary incompatibilities (though mima says it's fine as far as I can see).

**Comment by oleg-smith** [comment]
@jrudolph what do you mean by that, specifically? How is case class extended except addind case class methods of previous versions?

**Comment by oleg-smith** [comment]
@raboof Do you think it can be merged

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 3 above)
Comment (ktoso) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: Using an default value is a bad ideas as with those it becomes even harder to keep any bin compat as one later loses the capability to do overloads properly. Please remove the default value, we need to provide methods that achieve the same effect, which I think have been added.

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala [hunk 4]
```diff
@@ -154,7 +148,46 @@ object Unmarshaller
...
+  }
+
   object UnsupportedContentTypeException {
-    def apply(supported: ContentTypeRange*): UnsupportedContentTypeException = UnsupportedContentTypeException(Set(supported: _*))
+
+    @deprecated("for binary compatibility")
+    def apply(supported: ContentTypeRange*): UnsupportedContentTypeException =
+      new UnsupportedContentTypeException(Set(supported: _*), None)
+
+    @deprecated("for binary compatibility")
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: yes, for compatibility but not really deprecated, the apply is perfectly fine as is.

**Comment by ktoso** [comment]
I can try pushing this over the finish line, the binary-compatibility aspect is pretty tricky in this one. You've done a good job about it @oleg-smith but I'm still worried about a few things, the default param and the way we added the field etc.

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 1 above)
Comment (raboof) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: We would indeed have to add some additional explicit methods (like `equals`), but it would avoid adding yet another automatically-generated `copy` that we'd have to handle when we change this type... OK though.

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 3 above)
Comment (raboof) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: Perhaps we could make the naming more descriptive here? `actual`?

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 4 above)
Comment (oleg-smith) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: should i remove? as i was asked to add `deprecated` directly before

**Comment by lightbend-cla-validator** [comment]
Hi @oleg-smith,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 3 above)
Comment (oleg-smith) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: @ktoso removed the default value

**Comment by oleg-smith** [comment]
Signed the CLA

**Comment by oleg-smith** [comment]
@raboof @ktoso Does it look ok?

**Comment by raboof** [comment]
(verified you have indeed signed the CLA, not sure why the github bot isn't picking it up yet)

**Comment by oleg-smith** [comment]
@ktoso could you check when you have a chance?

**Comment by raboof** [comment]
PLS BUILD

**Comment by jlprat** [comment]
I'm quite busy at the moment, I'll try to take a look, but don't wait for me :)

**Comment by oleg-smith** [comment]
@raboof Providing the other guys are really busy, can we merge it? :)

File: akka-http-core/src/test/scala/akka/http/scaladsl/GracefulTerminationSpec.scala [hunk 5]
```diff
@@ -122,7 +122,7 @@ class GracefulTerminationSpec extends WordSpec with Matchers with BeforeAndAfter
 
     "in-flight request responses should include Connection: close and connection should be closed" in new TestSetup {
       val r1 = makeRequest() // establish connection
-      val time: FiniteDuration = 3.seconds
+      val time: FiniteDuration = 6.seconds
```
Comment (johanandren) [inline akka-http-core/src/test/scala/akka/http/scaladsl/GracefulTerminationSpec.scala]: This is unrelated, revert.

File: akka-http/src/main/scala/akka/http/scaladsl/common/StrictForm.scala [hunk 6]
```diff
@@ -125,10 +126,10 @@ object StrictForm {
           }
 
         tryUnmarshalToQueryForm.fast.recoverWith {
-          case Unmarshaller.UnsupportedContentTypeException(supported1) ⇒
+          case UnsupportedContentTypeException(supported1, _) ⇒
```
Comment (johanandren) [inline akka-http/src/main/scala/akka/http/scaladsl/common/StrictForm.scala]: if this is required we still have a binary compatibility problem, I think we need the single param unapply or else it could potentially break usage like the old one.

File: akka-http/src/main/scala/akka/http/scaladsl/server/Rejection.scala [hunk 7]
```diff
@@ -111,10 +110,39 @@ final case class InvalidOriginRejection(allowedOrigins: immutable.Seq[SHttpOrigi
...
+final case class UnsupportedRequestContentTypeRejection(
+  supported:   Set[ContentTypeRange],
+  contentType: Option[ContentType]   = None)
+  extends jserver.UnsupportedRequestContentTypeRejection
+  with Rejection {
+
   override def getSupported: java.util.Set[model.ContentTypeRange] =
     scala.collection.mutable.Set(supported.map(_.asJava).toVector: _*).asJava // TODO optimise
+
+  @deprecated("for binary compatibility")
```
Comment (johanandren) [inline akka-http/src/main/scala/akka/http/scaladsl/server/Rejection.scala]: All these should have a version as well, so `@deprecated("Something something", since = "10.1.6")`, they should also rather mention what to use instead. So, in this case "Use the two parameter constructor instead".

For the copy I guess the alternative now is to create a new object manually, so same there.

File: akka-http/src/main/scala/akka/http/scaladsl/server/Rejection.scala [hunk 8]
```diff
@@ -111,10 +110,39 @@ final case class InvalidOriginRejection(allowedOrigins: immutable.Seq[SHttpOrigi
...
+  def copy$default$1(supported: Set[ContentTypeRange]) =
+    new UnsupportedRequestContentTypeRejection(supported, this.contentType)
+
+  @deprecated("for binary compatibility")
+  def copy(
+    supported:   Set[ContentTypeRange] = this.supported,
+    contentType: Option[ContentType]   = this.contentType) =
+    UnsupportedRequestContentTypeRejection(supported, contentType)
```
Comment (johanandren) [inline akka-http/src/main/scala/akka/http/scaladsl/server/Rejection.scala]: This is the new two-param copy so it shouldn't be deprecated?

File: akka-http/src/main/scala/akka/http/javadsl/server/Rejections.scala [hunk 9]
```diff
@@ -376,8 +376,18 @@ object Rejections {
...
     s.MalformedHeaderRejection(headerName, errorMsg, cause.asScala)
 
+  def unsupportedRequestContentType(
+    supported:   java.lang.Iterable[MediaType],
+    contentType: Optional[ContentType]): UnsupportedRequestContentTypeRejection =
+    s.UnsupportedRequestContentTypeRejection(
+      supported = supported.asScala.map(m ⇒ scaladsl.model.ContentTypeRange(m.asScala)).toSet,
+      contentType = contentType.asScala.map(_.asScala))
+
+  // for backwards compatibility
```
Comment (johanandren) [inline akka-http/src/main/scala/akka/http/javadsl/server/Rejections.scala]: should be marked deprecated as well

File: akka-http/src/main/scala/akka/http/scaladsl/server/RejectionHandler.scala [hunk 10]
```diff
@@ -257,7 +257,8 @@ object RejectionHandler {
       }
       .handleAll[UnsupportedRequestContentTypeRejection] { rejections ⇒
         val supported = rejections.flatMap(_.supported).mkString(" or ")
-        rejectRequestEntityAndComplete((UnsupportedMediaType, "The request's Content-Type is not supported. Expected:\n" + supported))
+        val unsupportedContentType = rejections.find(_.contentType.isDefined).map(_.contentType)
```
Comment (johanandren) [inline akka-http/src/main/scala/akka/http/scaladsl/server/RejectionHandler.scala]: This will be an `Option[Option[ContentType]]` now. I think you can just do a `rejections.flatMap(_.contentType)` instead.

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala [hunk 11]
```diff
@@ -47,38 +49,44 @@ object Unmarshaller
...
 
   /**
-   * Helper for creating a "super-unmarshaller" from a sequence of "sub-unmarshallers", which are tried
-   * in the given order. The first successful unmarshalling of a "sub-unmarshallers" is the one produced by the
-   * "super-unmarshaller".
-   */
+    * Helper for creating a "super-unmarshaller" from a sequence of "sub-unmarshallers", which are tried
+    * in the given order. The first successful unmarshalling of a "sub-unmarshallers" is the one produced by the
+    * "super-unmarshaller".
+    */
```
Comment (johanandren) [inline akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: This is not the scaladoc style we use, please revert

File: akka-http-core/src/test/scala/akka/http/scaladsl/GracefulTerminationSpec.scala (hunk 5 above)
Comment (oleg-smith) [reply akka-http-core/src/test/scala/akka/http/scaladsl/GracefulTerminationSpec.scala]: This test fails due to lack of time on Jenkins. How can I make the unrelated test to pass otherwise?

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 11 above)
Comment (oleg-smith) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: doesn't `validatePullRequest` format code properly?

File: akka-http-core/src/test/scala/akka/http/scaladsl/GracefulTerminationSpec.scala (hunk 5 above)
Comment (johanandren) [reply akka-http-core/src/test/scala/akka/http/scaladsl/GracefulTerminationSpec.scala]: Then it should fail for all other PR validation and our nightlies on Jenkins as well, which I'm pretty sure it doesn't.

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 11 above)
Comment (johanandren) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: Could be that it doesn't reformat scaladoc, not sure.

File: akka-http-core/src/test/scala/akka/http/scaladsl/GracefulTerminationSpec.scala (hunk 5 above)
Comment (oleg-smith) [reply akka-http-core/src/test/scala/akka/http/scaladsl/GracefulTerminationSpec.scala]: No, it should not - it can fail from time to time depending on agent.

**Comment by oleg-smith** [comment]
@raboof yeah, will do

File: akka-http/src/main/scala/akka/http/scaladsl/common/StrictForm.scala (hunk 6 above)
Comment (oleg-smith) [reply akka-http/src/main/scala/akka/http/scaladsl/common/StrictForm.scala]: isn't binary compatibility checked automatically so nobody guesses about it?

File: akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala (hunk 11 above)
Comment (raboof) [reply akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/Unmarshaller.scala]: I am also surprised that scalariform doesn't update this indentation, but I just checked and indeed it doesn't.

File: akka-http/src/main/scala/akka/http/scaladsl/common/StrictForm.scala (hunk 6 above)
Comment (raboof) [reply akka-http/src/main/scala/akka/http/scaladsl/common/StrictForm.scala]: Unfortunately there are some cases that are not yet caught by MiMa, and this is one of them: https://github.com/lightbend/migration-manager/issues/40

**Comment by raboof** [comment]
(closing due to inactivity, feel free to reopen when someone wants to resume this work!)
